File: Ruby/Euler021.rb
#!/usr/bin/env ruby

def main
  target = 10000
  proper_divisors_sums = {}
  amicable_sum = 0

  (1..target).each { |i|
    if !proper_divisors_sums.key? i
      proper_divisors_sums[i] = get_proper_divisors_sum(i)
    end
    divisors_sum = proper_divisors_sums[i]

    if !proper_divisors_sums.key? divisors_sum
      proper_divisors_sums[divisors_sum] = get_proper_divisors_sum(divisors_sum)
    end
    other_divisors_sum = proper_divisors_sums[divisors_sum]

    if other_divisors_sum == i && i != divisors_sum
      amicable_sum += i
    end
  }
  puts amicable_sum
end

def get_proper_divisors_sum(n)
  limit = Math.sqrt(n).ceil.to_i
  nums = (2..limit).filter { |x| n % x == 0 }
  nums += nums.map { |x| n / x }
  return nums.uniq.sum + 1
end

if __FILE__ == $0
  main
end

File: Node/Euler021.js
#!/usr/bin/env node

function main() {
  const target = 10000;
  let properDivisorsSums = {};
  let amicableSum = 0;

  for (let i = 1; i < target; i++) {
    let divisorsSum = properDivisorsSums[i];
    if (divisorsSum === undefined) {
      divisorsSum = getProperDivisorsSum(i);
      properDivisorsSums[i] = divisorsSum;
    }

    let otherDivisorsSum = properDivisorsSums[divisorsSum];
    if (otherDivisorsSum === undefined) {
      otherDivisorsSum = getProperDivisorsSum(divisorsSum);
      properDivisorsSums[divisorsSum] = otherDivisorsSum;
    }

    if (otherDivisorsSum == i && i != divisorsSum) {
      amicableSum += i;
    }
  }

  console.log(amicableSum);
}

function getProperDivisorsSum(n) {
  let sum = 1;
  let limit = Math.ceil(Math.sqrt(n));

  for (let i = 2; i <= limit; i++) {
    if (n % i == 0) {
      sum += i;
      let otherDivisor = n / i;
      if (otherDivisor != i) {
        sum += otherDivisor;
      }
    }
  }

  return sum
}

if (require.main == module) {
  main();
}
